src/seccomp/src/bin.rs

```diff
@@ -1,8 +1,568 @@
+
+#[derive(Debug)]
+pub enum Error {
+    Bincode(BincodeError),
+    FileOpen(PathBuf, io::Error),
+    FileFormat(FilterFormatError),
+    InvalidArch,
+    JSON(JSONError),
+    MissingInputFile,
+}
+
+type Result<T> = std::result::Result<T, Error>;
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use self::Error::*;
+
+        match *self {
+            Bincode(ref err) => write!(f, "Bincode (de)serialization failed: {}", err),
+            FileFormat(ref err) => write!(f, "{}", err),
+            FileOpen(ref path, ref err) => write!(
+                f,
+                "{}",
+                format!("Failed to open file {:?}: {}", path, err).replace("\"", "")
+            ),
+            InvalidArch => write!(f, "Invalid arch"),
+            JSON(ref err) => write!(f, "Error parsing JSON: {}", err),
+            MissingInputFile => write!(f, "Missing input file"),
+        }
+    }
+}
+
+#[derive(Debug, PartialEq)]
+struct Arguments {
+    input_file: String,
+    output_file: String,
+    target_arch: String,
+}
+
+fn build_arg_parser() -> ArgParser<'static> {
+    ArgParser::new()
+        .arg(
+            Argument::new("input_file")
+                .required(true)
+                .takes_value(true)
+                .help("File path of the JSON input."),
+        )
+        .arg(
+            Argument::new("output_file")
+                .required(false)
+                .takes_value(true)
+                .default_value("bpf.out")
+                .help("Optional path of the output file."),
+        )
+        .arg(
+            Argument::new("target_arch")
+                .required(true)
+                .takes_value(true)
+                .help("ARCH of the CPU that will be used to run the BPF program. One of x86_64/aarch64"),
+        )
+}
+
+fn get_argument_values(arguments: &ArgumentsBag) -> Result<Arguments> {
+    let target_arch = arguments.value_as_string("target_arch").and_then(|val| {
+        if !SUPPORTED_ARCHES.contains(&val.as_ref()) {
+            None
+        } else {
+            Some(val)
+        }
+    });
+
+    if target_arch.is_none() {
+        return Err(Error::InvalidArch);
+    }
+
+    let input_file = arguments.value_as_string("input_file");
+    if input_file.is_none() {
+        return Err(Error::MissingInputFile);
+    }
+
+    Ok(Arguments {
+        target_arch: target_arch.unwrap(),
+        input_file: input_file.unwrap(),
+        // Safe to unwrap because it has a default value
+        output_file: arguments.value_as_string("output_file").unwrap(),
+    })
+}
+
+fn parse_json(reader: &mut dyn Read) -> Result<HashMap<String, Filter>> {
+    serde_json::from_reader(reader).map_err(Error::JSON)
+}
+
+fn compile(args: &Arguments) -> Result<()> {
+    let input_file = File::open(&args.input_file)
+        .map_err(|err| Error::FileOpen(PathBuf::from(&args.input_file), err))?;
+    let mut input_reader = BufReader::new(input_file);
```

Comment:
Actually, I don't know for sure how many reads the program makes to the input file. This is in `serde_json`'s hands.
I am assuming that it does more than one `read` calls, to not bring into memory the entire file at once (the file could be huuge).
It would have been clear that it does only one `read` if seccompiler used `deserialize_str`. However, this is precisely why I chose to use `deserialize_from` (which works with Read objects and likely does more read actions)

Reply:
Update: 
https://docs.serde.rs/serde_json/fn.from_reader.html
`serde_json` actually recommends using BufReader for File objects.

This should also take care of OOM. WDYT?

Reply:
`from_reader` actually passes the burden of reading stuff to the `Read`er, it doesn't handle the actual reading ([here](https://github.com/serde-rs/json/blob/3d8b046d0247da886a60de0d1b1aa8ba1524cfe5/src/de.rs#L69) to [here](https://github.com/serde-rs/json/blob/3d8b046d0247da886a60de0d1b1aa8ba1524cfe5/src/read.rs#L190)); which is why I was raising the `BufReader` question, as it's indeed preferable (and recommended by the documentation) for *small* reads, but that doesn't apply to this use case, where you need to read the whole thing into memory for `serde_json` to convert it.

Reply:
Running seccompiler with BufReader under strace, on a file with ~600 lines of JSON, I see three calls to `read`.
In contrast, running without BufReader, seccompiler makes a call to `read` for **every single byte**.

I have also done a small benchmark test based on the `time` command, and for a ~600 lines JSON file, the BufReader approach is about ~0.01s faster. This is definitely not much, and given the fact that `seccompiler` is not really performance constrained, it really does not matter that much.

However, given the huuuge amount of syscalls, I think it would be best to keep BufReader